src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -0,0 +1,524 @@
+    input.set_size(inputElements, 1);
+
+    MultiplyConstant<> module(3.0);
+
+    double error = JacobianTest(module, input);
+    BOOST_REQUIRE_LE(error, 1e-5);
+  }
+}
+
+/**
+ * Jacobian HardTanH module test.
+ */
+BOOST_AUTO_TEST_CASE(JacobianHardTanHLayerTest)
+{
+  for (size_t i = 0; i < 5; i++)
+  {
+    const size_t inputElements = math::RandInt(2, 1000);
+
+    arma::mat input;
+    input.set_size(inputElements, 1);
+
+    HardTanH<> module;
+
+    double error = JacobianTest(module, input);
+    BOOST_REQUIRE_LE(error, 1e-5);
+  }
+}
+
+/**
+ * Simple select module test.
+ */
+BOOST_AUTO_TEST_CASE(SimpleSelectLayerTest)
+{
+  arma::mat outputA, outputB, input, delta;
+
+  input = arma::ones(10, 5);
+  for (size_t i = 0; i < input.n_cols; ++i)
+  {
+    input.col(i) *= i;
+  }
+
+  // Test the Forward function.
+  Select<> moduleA(3);
+  moduleA.Forward(std::move(input), std::move(outputA));
+  BOOST_REQUIRE_EQUAL(30, arma::accu(outputA));
+
+  // Test the Forward function.
+  Select<> moduleB(3, 5);
+  moduleB.Forward(std::move(input), std::move(outputB));
+  BOOST_REQUIRE_EQUAL(15, arma::accu(outputB));
+
+  // Test the Backward function.
+  moduleA.Backward(std::move(input), std::move(outputA), std::move(delta));
+  BOOST_REQUIRE_EQUAL(30, arma::accu(delta));
+
+  // Test the Backward function.
+  moduleB.Backward(std::move(input), std::move(outputA), std::move(delta));
+  BOOST_REQUIRE_EQUAL(15, arma::accu(delta));
+}
+
+/**
+ * Simple join module test.
+ */
+BOOST_AUTO_TEST_CASE(SimpleJoinLayerTest)
+{
+  arma::mat output, input, delta;
+  input = arma::ones(10, 5);
+
+  // Test the Forward function.
+  Join<> module;
+  module.Forward(std::move(input), std::move(output));
+  BOOST_REQUIRE_EQUAL(50, arma::accu(output));
+
+  bool b = output.n_rows == 1 || output.n_cols == 1;
+  BOOST_REQUIRE_EQUAL(b, true);
+
+  // Test the Backward function.
+  module.Backward(std::move(input), std::move(output), std::move(delta));
+  BOOST_REQUIRE_EQUAL(50, arma::accu(delta));
+
+  b = delta.n_rows == input.n_rows && input.n_cols;
+  BOOST_REQUIRE_EQUAL(b, true);
+}
+
+/**
+ * Simple add merge module test.
+ */
+BOOST_AUTO_TEST_CASE(SimpleAddMergeLayerTest)
+{
+  arma::mat output, input, delta;
+  input = arma::ones(10, 1);
+
+  for (size_t i = 0; i < 5; ++i)
+  {
+    AddMerge<> module;
+    const size_t numMergeModules = math::RandInt(2, 10);
+    for (size_t m = 0; m < numMergeModules; ++m)
+    {
+      IdentityLayer<> identityLayer;
+      identityLayer.Forward(std::move(input),
+          std::move(identityLayer.OutputParameter()));
+
+      module.Add(identityLayer);
+    }
+
+    // Test the Forward function.
+    module.Forward(std::move(input), std::move(output));
+    BOOST_REQUIRE_EQUAL(10 * numMergeModules, arma::accu(output));
+
+    // Test the Backward function.
+    module.Backward(std::move(input), std::move(output), std::move(delta));
+    BOOST_REQUIRE_EQUAL(arma::accu(output), arma::accu(delta));
+  }
+}
```

Comment:
Yeah, implementing more tests is definitely on my list, I think I used at least every layer in one test. Like there is no test for the base layer in ``ann_layer_test``, since each activation function is tested in the ``activation_functions_test`` or the mean pooling layer is tested in ``pooling_test``.